jobs-forwarder/internal/forwarder/jobsforwarder.go

```diff
@@ -0,0 +1,150 @@
+package forwarder
+
+import (
+	"context"
+	"encoding/json"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+
+	pulsarType "github.com/apache/pulsar-client-go/pulsar"
+	"github.com/rudderlabs/rudder-go-kit/config"
+	"github.com/rudderlabs/rudder-go-kit/logger"
+	backendconfig "github.com/rudderlabs/rudder-server/backend-config"
+	"github.com/rudderlabs/rudder-server/internal/pulsar"
+	"github.com/rudderlabs/rudder-server/jobs-forwarder/internal/schematransformer"
+	"github.com/rudderlabs/rudder-server/jobsdb"
+	"github.com/rudderlabs/rudder-server/utils/misc"
+)
+
+type JobsForwarder struct {
+	BaseForwarder
+	pulsarProducer pulsar.ProducerAdapter
+	backendConfig  backendconfig.BackendConfig
+	transformer    schematransformer.Transformer
+	retryAttempts  int
+	key            string
+}
+
+func NewJobsForwarder(ctx context.Context, g *errgroup.Group, schemaDB jobsdb.JobsDB, client *pulsar.Client, config *config.Config, backendConfig backendconfig.BackendConfig, log logger.Logger) (*JobsForwarder, error) {
+	forwarder := JobsForwarder{
+		retryAttempts: config.GetInt("JobsForwarder.retryAttempts", 3),
+		key:           config.GetString("JobsForwarder.key", "event-schema"),
+	}
+	forwarder.LoadMetaData(ctx, g, schemaDB, log, config)
+	producer, err := client.NewProducer(config)
+	if err != nil {
+		return nil, err
+	}
+	forwarder.pulsarProducer = producer
+	forwarder.backendConfig = backendConfig
+	forwarder.transformer = schematransformer.New(ctx, g, backendConfig, config)
+	return &forwarder, nil
+}
+
+func (jf *JobsForwarder) Start() {
+	var sleepTime time.Duration
+	jf.transformer.Start()
+	jf.g.Go(misc.WithBugsnag(func() error {
+		for {
+			select {
+			case <-jf.ctx.Done():
+				return nil
+			case <-time.After(sleepTime):
+				jobs, limitReached, err := jf.GetJobs(jf.ctx)
+				if err != nil {
+					return err
+				}
+				filteredJobs, statusList := jf.filterJobs(jobs)
+				for _, job := range filteredJobs {
+					transformedBytes, orderingKey, err := jf.transformer.Transform(job)
+					if err != nil {
+						statusList = append(statusList, &jobsdb.JobStatusT{
+							JobID:         job.JobID,
+							AttemptNum:    job.LastJobStatus.AttemptNum + 1,
+							JobState:      jobsdb.Failed.State,
+							ExecTime:      time.Now(),
+							RetryTime:     time.Now(),
+							ErrorCode:     "500",
+							Parameters:    []byte{},
+							ErrorResponse: json.RawMessage(err.Error()),
+						})
+						continue
+					}
+					statusFunc := func(_ pulsarType.MessageID, _ *pulsarType.ProducerMessage, err error) {
```

Comment:
We should retry failed messages, with some retries and backoff, until all messages of the batch have been sent. After failing all the retries we can abort and then pick the next batch. The retry period though should be really big, e.g. a few hours.

It is not expected for messages to fail at this stage, only if there is a technical issue (network, pulsar unavailability etc.)